src/backend/staking/main.mo

```diff
@@ -375,6 +367,217 @@ actor class DoxaStaking() = this {
+
+			// Calculate and store rewards for all active stakes
+			await distributeWeeklyRewards(currentWeekNumber, weeklyReward);
+		};
+	};
+
+	// Function to distribute rewards to all stakers
+	private func distributeWeeklyRewards(weekNumber : Nat, totalReward : Nat) : async () {
+		for ((principal, stakeIds) in Map.entries(userStakes)) {
+			var userTotalWeight : Nat = 0;
+
+			// Calculate user's total reward across all stakes
+			var userTotalReward : Nat = 0;
+			for (stakeId in stakeIds.vals()) {
+				switch (Map.get(stakes, nhash, stakeId)) {
+					case (?stake) {
+						let metrics = await calculateUserStakeMatric(stakeId, principal);
+						switch (metrics) {
+							case (#ok(m)) {
+								userTotalReward += m.finalReward;
+							};
+							case (#err(_)) {};
+						};
+					};
+					case (null) {};
+				};
+			};
+
+			// Use total reward from metrics
+			let userReward = userTotalReward;
+
+			// Store user's reward for this week
+			switch (Map.get(userRewardMap, phash, principal)) {
+				case (?weekMap) {
+					Map.set(weekMap, nhash, weekNumber, userReward);
+				};
+				case (null) {
+					let newWeekMap = Map.new<Nat, Nat>();
+					Map.set(newWeekMap, nhash, weekNumber, userReward);
+					Map.set(userRewardMap, phash, principal, newWeekMap);
+				};
+			};
+		};
+	};
+
+	// Add new enum for harvest action type
+	public type HarvestAction = {
+		#Withdraw; // Withdraw rewards to wallet
+		#Compound; // Add rewards to existing stake
+	};
+
+	// Modified harvest function with compound option
+	public shared ({ caller }) func harvestRewards(stakeId : Types.StakeId, action : HarvestAction) : async Result.Result<(), Text> {
+		// Verify stake belongs to caller
+		let userStakeIds = switch (Map.get(userStakes, phash, caller)) {
+			case (null) return #err("No stakes found for user");
+			case (?ids) ids;
+		};
+
+		// Check if stakeId exists in user's stakes
+		let validStakeId = Array.find<Types.StakeId>(userStakeIds, func(id) { id == stakeId });
+		if (validStakeId == null) {
+			return #err("This stake ID does not belong to caller");
+		};
+
+		// Get stake details
+		let stake = switch (Map.get(stakes, nhash, stakeId)) {
+			case (null) return #err("Stake not found");
+			case (?s) {
+				if (s.staker != caller) return #err("Not authorized to access this stake");
+				s;
+			};
+		};
+
+		switch (Map.get(userRewardMap, phash, caller)) {
+			case (?weekMap) {
+				var totalReward : Nat = 0;
+
+				// Sum up all unclaimed rewards
+				for ((weekNum, reward) in Map.entries(weekMap)) {
+					totalReward += reward;
+				};
+
+				if (totalReward == 0) {
+					return #err("No rewards available to harvest");
+				};
+
+				switch (action) {
+					case (#Withdraw) {
+						// Original withdraw logic
+						let transferResult = await USDx.icrc1_transfer({
+							to = { owner = caller; subaccount = null };
+							amount = totalReward;
+							fee = null;
+							memo = null;
+							from_subaccount = null;
```

Comment:
We need to specify a dedicated subaccount of the staking canister for fee storage. 
This will help in better tracking and management of fees separately